storage/SMAIndexSubBlock.cpp

```diff
@@ -0,0 +1,670 @@
+
+  // Clearing min/max will free any out of line data.
+  if (entry->min_entry_ref.valid) {
+    entry->min_entry_ref.value.clear();
+    entry->min_entry_ref.valid = false;
+  }
+  if (entry->max_entry_ref.valid) {
+    entry->max_entry_ref.value.clear();
+    entry->max_entry_ref.valid = false;
+  }
+  sma_internal::initializeTypedValueForSum(entry);
+}
+
+void SMAIndexSubBlock::resetEntries() {
+  freeOutOfLineData();
+
+  for (std::size_t indexed_attribute_num = 0;
+       indexed_attribute_num < indexed_attributes_;
+       ++indexed_attribute_num) {
+    const attribute_id attribute = description_.GetExtension(
+        SMAIndexSubBlockDescription::indexed_attribute_id,
+        indexed_attribute_num);
+
+    const Type &attribute_type = tuple_store_.getRelation().getAttributeById(attribute)->getType();
+    resetEntry(entries_ + indexed_attribute_num, attribute, attribute_type);
+  }
+}
+
+void SMAIndexSubBlock::freeOutOfLineData() {
+  // For each entry, clear the min and max typed values (frees out of line data).
+  for (std::size_t indexed_attribute_num = 0;
+       indexed_attribute_num < indexed_attributes_;
+       ++indexed_attribute_num) {
+    SMAEntry &entry = entries_[indexed_attribute_num];
+    if (!TypedValue::RepresentedInline(entry.type_id)) {
+      if (entry.min_entry_ref.valid) {
+        entry.min_entry_ref.value.clear();
+      }
+      if (entry.max_entry_ref.valid) {
+        entry.max_entry_ref.value.clear();
+      }
+    }
+  }
+}
+
+bool SMAIndexSubBlock::DescriptionIsValid(const CatalogRelationSchema &relation,
+                                          const IndexSubBlockDescription &description) {
+  if (!description.IsInitialized()) {
+    return false;
+  }
+  if (description.sub_block_type() != IndexSubBlockDescription::SMA) {
+    return false;
+  }
+  // Must have at least one indexed attribute.
+  if (description.ExtensionSize(SMAIndexSubBlockDescription::indexed_attribute_id) == 0) {
+    return false;
+  }
+  // Check that all key attributes exist.
+  for (int indexed_attribute_num = 0;
+       indexed_attribute_num < description.ExtensionSize(SMAIndexSubBlockDescription::indexed_attribute_id);
+       ++indexed_attribute_num) {
+    attribute_id indexed_attribute_id = description.GetExtension(
+        SMAIndexSubBlockDescription::indexed_attribute_id, indexed_attribute_num);
+    if (!relation.hasAttributeWithId(indexed_attribute_id)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+std::size_t SMAIndexSubBlock::EstimateBytesPerTuple(
+    const CatalogRelationSchema &relation,
+    const IndexSubBlockDescription &description) {
+  // TODO(marc): Returning 1 almost always gives too much space. An enhancement
+  // would be to inform the storage subblock that this index uses space w.r.t.
+  // # of attributes being indexed, not tuples.
+  return 1;
+}
+
+bool SMAIndexSubBlock::bulkAddEntries(const TupleIdSequence &tuples) {
+  header_->index_consistent = false;
```

Comment:
We might want to talk about this one in person because there performance issues associated with updating the index during point inserts/deletes versus the method which I was planning to use which was to mark the index inconsistent and then rebuild it when necessary.